src/game/Utils/Cinematics.cc

```diff
@@ -7,200 +7,313 @@
+		sf->file_in_memory = MALLOCN(unsigned char, bytes);
+		FileRead(file, sf->file_in_memory, bytes);
+
+		// open with smacker
+		sf->smacker = smk_open_memory(sf->file_in_memory, bytes);
+		if (sf->smacker == nullptr) throw new std::runtime_error("smk_open_memory failed");
+		sf->flags |= SMK_FLIC_OPEN;
+		return sf;
 	}
+	catch (const std::runtime_error& ex)
+	{
 
-	AutoSGPFile file(GCM->openGameResForReading(filename));
+		SLOGE("Failed to open '%s': %s", filename, ex.what());
+		if (sf != nullptr) SmkCloseFlic(sf);
+		return nullptr;
+	}
+	return sf;
+}
 
-	//FILE* const f = GetRealFileHandleFromFileManFileHandle(file);
 
-	//printf("File Size: %u\n", FileGetSize(file));
-	// Allocate a Smacker buffer for video decompression
-	/*
-	sf->SmackBuffer = SmackBufferOpen(SMACKAUTOBLIT, SCREEN_WIDTH, SCREEN_HEIGHT, 0, 0);
-	if (sf->SmackBuffer == NULL)
+void SmkCloseFlic(SMKFLIC* const sf)
+{
+	Assert(sf != nullptr);
+	FOR_EACH(UINT32, sound, sf->sounds)
+	{
+		if (*sound != NO_SAMPLE)
+		{
+			SoundStop(*sound);
+			*sound = NO_SAMPLE;
+		}
+	}
+	if (sf->smacker != nullptr)
 	{
-		SLOGE("Can't allocate a Smacker decompression buffer");
-		return NULL;
+		smk_close(sf->smacker);
+		sf->smacker = nullptr;
 	}
-	*/
-	sf->SmackerObject = SmackOpen( file , SMACKFILEHANDLE | SMACKTRACKS, SMACKAUTOEXTRA);
-	if (!sf->SmackerObject)
+	if (sf->file_in_memory != nullptr)
 	{
-		SLOGE("Smacker won't open the SMK file");
-		return NULL;
+		MemFree(sf->file_in_memory);
+		sf->file_in_memory = nullptr;
 	}
-
-	// Make sure we have a video surface
-	SmkSetupVideo();
-
-	sf->hFileHandle  = file.Release();
-	sf->uiFlags     |= SMK_FLIC_OPEN;
-	return sf;
+	sf->flags = 0;
 }
-catch (...) { return 0; }
 
 
-void SmkCloseFlic(SMKFLIC* const sf)
+static SMKFLIC* SmkGetFreeFlic(void)
 {
-	SmackClose(sf->SmackerObject);
-	//FileClose(sf->hFileHandle);
-	// reenable for blitting SmackBufferClose(sf->SmackBuffer);
-	memset(sf, 0, sizeof(*sf));
+	FOR_EACH(SMKFLIC, sf, gSmkList)
+	{
+		if (!(sf->flags & SMK_FLIC_OPEN)) return sf;
+	}
+	return nullptr;
 }
 
 
-static SMKFLIC* SmkGetFreeFlic(void)
+static void SmkSkipFrames(SMKFLIC* sf)
 {
-	FOR_EACH(SMKFLIC, i, SmkList)
+	// get target frame
+	UINT32 milliseconds = SDL_GetTicks() - sf->start_tick;
+	UINT32 frame_no = static_cast<UINT32>(milliseconds / sf->milliseconds_per_frame);
+
+	// skip until the target frame (video repeats if there is a ring frame)
+	while (sf->status != SMK_ERROR && sf->status != SMK_DONE && sf->frame_no != frame_no)
 	{
-		if (!(i->uiFlags & SMK_FLIC_OPEN)) return i;
+		sf->status = smk_next(sf->smacker);
+		sf->frame_no++;
 	}
-	return NULL;
 }
 
 
-static void SmkSetupVideo(void)
+static void SmkBlitVideoFrame(SMKFLIC* const sf, SGPVSurface* surface)
 {
-	UINT32 red;
-	UINT32 green;
-	UINT32 blue;
-	GetPrimaryRGBDistributionMasks(&red, &green, &blue);
-	if (red == 0xf800 && green == 0x07e0 && blue == 0x001f)
+	// get frame (source)
+	// TODO handle flags SMK_FLAG_Y_* (I need a sample of each case)
```

Comment:
Is this TODO still relevant?

Reply:
I would say yes, classic ja2 does not use the other modes, but mods might.
It's not critical, the video will just have an unexpected height.